<!-- Launchfile to start robot motion-->
<launch>
    <!-- Arguments -->
    <arg name="cmd_src" default="none" doc="The source of the cmd_vel command"/>
    <arg name="robot" default="nusim" doc="Whether the robot is running in simulation or real world"/>
    <arg name="use_rviz" default="true" doc="Set true to display rviz"/>
    <arg name="noiseless" default="false" doc="Set true to run robot without noise"/>
    <arg name="slammin" default="false" doc="Set true if running with slam. This will create a static transform from world to map"/>

    <!-- Parse cmd_src -->
    <group if="$(eval cmd_src == 'circle')">
        <node name="circle" pkg="nuturtle_control" type="circle"/>
    </group>
    
    <group if="$(eval cmd_src == 'teleop')">
        <node pkg="turtlebot3_teleop" type="turtlebot3_teleop_key" name="turtlebot3_teleop_keyboard"  output="screen"/>
    </group>


    <!-- Parse robot -->
    <!-- Nusim running-->
    <group if="$(eval robot == 'nusim')">
        <!--Nodes-->
        <node pkg="nuturtle_control" type="odometry" name="pure_odometry">
            <param name="body_id" value="blue-base_footprint"/>
            <param name="odom_id" value="world"/>
        </node>
        <node pkg="nuturtle_control" type="turtle_interface" name="turtle_interface"/>
        <!--Static Publisher-->
        <node pkg="tf2_ros" type="static_transform_publisher" name="world_odom_broadcaster" 
                args="0 0 0 0 0 0 world odom" unless="$(arg slammin)"/>
        <node pkg="tf2_ros" type="static_transform_publisher" name="world_map_broadcaster" 
                args="0 0 0 0 0 0 world map" if="$(arg slammin)"/>
        <!-- Nusim -->
        <rosparam command="load" file="$(find nusim)/config/basic_world.yaml"/>
        <rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml"/>
        <!-- remap nusim topics-->
        <remap from="red/sensor_data" to="sensor_data"/>
        <remap from="red/wheel_cmd" to="wheel_cmd"/>
        <param name="nusim/red_space" type="bool" value="false"/>
        <node pkg="nusim" type="nusim" name="nusim"/>

        <group if="$(eval noiseless == true)">
            <param name="nusim/dist/vel_mean" value="0"/>
            <param name="nusim/dist/vel_variance" value="0"/>
            <param name="nusim/dist/slip_min" value="1"/>
            <param name="nusim/dist/slip_max" value="1"/>
            <param name="nusim/sensor/basic_sensor_variance" value="0"/>
        </group>

        <!--Params for odometry-->
        <param name="wheel_left" value="blue-wheel_left_joint"/>
        <param name="wheel_right" value="blue-wheel_right_joint"/>
        
        <!--Robots-->
        <param name="red/robot_description" 
            command="xacro $(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro color:='red'"/>
        <param name="robot_description" 
            command="xacro $(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro color:='blue'"/>
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" ns='red'/>
        <node pkg="joint_state_publisher" type="joint_state_publisher" 
            name="joint_state_publisher" ns="red"/>
    </group>

    <!-- Localhost and turtlebot -->
    <group if="$(eval (robot != 'nusim') and (robot != 'none'))">
        <!--Nodes-->
        <node pkg="nuturtle_control" type="odometry" name="pure_odometry" machine="turtlebot">
            <param name="body_id" value="blue-base_footprint"/>
            <param name="odom_id" value="world"/>
        </node>

        <node pkg="nuturtle_control" type="turtle_interface" name="turtle_interface" machine="turtlebot"/>
        <!--Static Publisher-->
        <node pkg="tf2_ros" type="static_transform_publisher" name="world_odom_broadcaster" 
                args="0 0 0 0 0 0 world odom" unless="$(arg slammin)"/>
        <node pkg="tf2_ros" type="static_transform_publisher" name="world_map_broadcaster" 
                args="0 0 0 0 0 0 world map" if="$(arg slammin)"/>
                
        <!-- Params -->
        <rosparam command="load" file="$(find nusim)/config/basic_world.yaml"/>
        <rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml"/>
        <param name="wheel_left" value="blue-wheel_left_joint"/>
        <param name="wheel_right" value="blue-wheel_right_joint"/>
        <!--Robots-->
        <param name="robot_description" 
            command="xacro $(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro color:='blue'"/>
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
        <!--Includes-->
        <include file="$(find nuturtle_control)/launch/basic_remote.launch">
            <arg name="robot" value="$(arg robot)"/>
        </include>
    </group>

    <!-- None option-->
    <group if="$(eval robot == 'none')">
        <!-- Params -->
        <rosparam command="load" file="$(find nusim)/config/basic_world.yaml"/>
        <rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml"/>
        <param name="body_id" value="blue-base_footprint"/>
        <param name="wheel_left" value="blue-wheel_left_joint"/>
        <param name="wheel_right" value="blue-wheel_right_joint"/>
        <param name="odom_id" value="odom"/>

        <!--Nodes-->
        <node pkg="nuturtle_control" type="odometry" name="odometry"/>
        <node pkg="nuturtle_control" type="turtle_interface" name="turtle_interface"/>
    </group>

    <!-- Open Rviz-->
    <node pkg="rviz" type="rviz" name="rviz" required="true" if="$(arg use_rviz)"
        args="-d $(find nuturtle_control)/config/odom.rviz"/>

</launch>